searx/user_settings.py

```diff
@@ -0,0 +1,239 @@
+from searx import settings
+from searx.languages import language_codes
+from searx.autocomplete import backends as autocomplete_backends
+from searx import utils
+
+
+class UnknownSetting(Exception):
+    pass
+
+
+class InvalidSetting(Exception):
+    pass
+
+
+class ValueErrorInvalidSetting(InvalidSetting):
+    pass
+
+
+class TypeErrorInvalidSetting(InvalidSetting):
+    pass
+
+
+class UserSettingsBase(object):
+    defaults = None
+    _collection_fields = {}
+
+    def __init__(self):
+        self._settings = dict()
+        self._is_empty = True
+        if self.defaults is None:
+            self.defaults = dict()
+        self._load_defaults()
+
+    def __repr__(self):
+        return repr(self._settings)
+
+    def _load_defaults(self):
+        self.restore_from_cookies(self.defaults)
+
+    def empty(self):
+        return self._is_empty
+
+    def get(self, setting):
+        if setting in self._settings:
+            return self._settings[setting]
+        else:
+            raise UnknownSetting()
+
+    def verify(self, setting, value):
+        try:
+            if hasattr(self, "verify_%s" % setting) and \
+               not getattr(self, "verify_%s" % setting)(value):
+                raise ValueErrorInvalidSetting((setting, value))
+        except TypeError:
+            raise TypeErrorInvalidSetting((setting, value))
+
+    def validate(self, setting, value):
+        if hasattr(self, "validate_%s" % setting):
+            return getattr(self, "validate_%s" % setting)(value)
+        else:
+            self.verify(setting, value)
+            return value
+
+    def set(self, setting, value):
+        self._settings[setting] = self.validate(setting, value)
+        self._is_empty = False
+
+    def form_set(self, setting, value):
+        self.set(setting, self._deserialize(setting, value))
+
+    def override_from_form(self, form_data):
+        form_data = utils.parse_form(form_data, self._collection_fields)
+        for field, value in form_data.iteritems():
+            if field in self._collection_fields.values():
+                self.set(field, value)
+            else:
+                self.form_set(field, value)
+
+    def _serialize(self, setting, value):
+        if hasattr(self, "serialize_%s" % setting):
+            return getattr(self, "serialize_%s" % setting)(value)
+        else:
+            return str(value)
+
+    def _deserialize(self, setting, value):
+        if hasattr(self, "deserialize_%s" % setting):
+            return getattr(self, "deserialize_%s" % setting)(value)
+        else:
+            return value
+
+    def restore_from_cookies(self, cookies):
+        for setting, value in cookies.iteritems():
+            self.set(setting, self._deserialize(setting, value))
+
+    def save_to_cookies(self, cookies):
```

Comment:
Should simply return with the dict object (line 97). With this minor modification we can remove the `cookies` parameter to avoid side effects. The function name could be `get_as_cookies` or something like this.